src/main/java/org/datadog/jenkins/plugins/datadog/clients/HttpClient.java

```diff
@@ -0,0 +1,291 @@
+        };
+    }
+
+    private static <T> T executeSynchronously(Supplier<Request> requestSupplier, HttpRetryPolicy retryPolicy, Function<String, T> responseParser) throws InterruptedException, TimeoutException, ExecutionException {
+        while (true) {
+            ContentResponse response;
+            try {
+                Request request = requestSupplier.get();
+                response = request.send();
+            } catch (TimeoutException | ExecutionException e) {
+                if (retryPolicy.shouldRetry(null)) {
+                    Thread.sleep(retryPolicy.backoff());
+                    continue;
+                } else {
+                    throw e;
+                }
+            }
+
+            int status = response.getStatus();
+            if (status >= 200 && status < 300) {
+                if (responseParser == null) {
+                    return null;
+                }
+
+                String content = response.getContentAsString();
+                return responseParser.apply(content);
+
+            } else {
+                if (retryPolicy.shouldRetry(response)) {
+                    Thread.sleep(retryPolicy.backoff());
+                    continue;
+                }
+
+                String additionalHint;
+                switch (status) {
+                    case HttpStatus.FORBIDDEN_403:
+                        additionalHint = "API key might be invalid, please check your config";
+                        break;
+                    case HttpStatus.NOT_FOUND_404:
+                    case HttpStatus.BAD_REQUEST_400:
+                        additionalHint = "Request URL might be invalid, please check your config";
+                        break;
+                    default:
+                        additionalHint = "";
+                        break;
+                }
+
+                throw new ResponseProcessingException("Received erroneous response " + response + ". "  + additionalHint);
+            }
+        }
+    }
+
+    private static void executeAsynchronously(Supplier<Request> requestSupplier, HttpRetryPolicy retryPolicy) {
+        Request request = requestSupplier.get();
+        request.send(new ResponseListener(2 * 1024 * 1024, requestSupplier, retryPolicy));
+    }
+
+    private static final class ResponseListener extends BufferingResponseListener {
+        private final Supplier<Request> requestSupplier;
+        private final HttpRetryPolicy retryPolicy;
+
+        public ResponseListener(int maxLength, Supplier<Request> requestSupplier, HttpRetryPolicy retryPolicy) {
+            super(maxLength);
+            this.requestSupplier = requestSupplier;
+            this.retryPolicy = retryPolicy;
+        }
+
+        @Override
+        public void onComplete(Result result) {
+            try {
+                Throwable failure = result.getFailure();
+                if (failure != null) {
+                    if (retryPolicy.shouldRetry(null)) {
```

Comment:
why don't we check if we should retry with the response of the result here?

Reply:
Having a failure usually means there is no response, i.e. we failed before the response could be received.

Reply:
Gotcha, thanks! Is there any downside in first getting the response, and then checking `shouldRetry` with it? That way if the response is `null`, the behavior hasn't changed, but if there is a response, we can use the additional logic in `shouldRetry`?

Reply:
Agree, that would allow to simplify the code a bit. Changed it accordingly